api/v1beta2/authpolicy_types.go

```diff
@@ -18,16 +18,19 @@ type AuthSchemeSpec struct {
 	// Authentication configs.
 	// At least one config MUST evaluate to a valid identity object for the auth request to be successful.
 	// +optional
+	// +kubebuilder:validation:MaxProperties=14
```

Comment:
See https://github.com/kubernetes-sigs/gateway-api/issues/995#issuecomment-1787005682

Reply:
We are going with the hard limits for now. We hope soon to be able to remove or substantially increase these limits of maximum number of configs per AP/RLP, by redefining those GWAPI types (HTTPRouteMatch and related) into our code base without the CEL validation rules. After all, Kuadrant API does not need to re-impose validation on those fields if they have been validated once applied in the target GWAPI resources.

Reply:
Sorry, can you briefly explain what is this about? O read from the link

```
This would work fine until v0.7, but since validation rules based on [Common Expression Language (CEL)](https://github.com/google/cel-go) have been added to the HTTPRoute (https://github.com/kubernetes-sigs/gateway-api/pull/2253), the estimated cost limit exceeds the thresholds, and we have been struggling with having to set some hard constraints on the number of policy rules a user can declare.
```

What do you mean by ` the estimated cost limit exceeds the thresholds` ? The client upon creating an object is giving timeouts because k8s API server takes too long to validate the resource?

Reply:
ping @guicassolato

Reply:
I can try to provide some info here @eguzki.

When you submit a CRD to the kube apiserver it makes an estimate of how costly the CEL validation rules are, so that when a CR of this type is submitted to the cluster it can evaluate it. There is a cost limit for this validation, and so if the estimated cost exceeds the limit the CRD is rejected, as the kube apiserver determines it would be too costly for it to perform validations on CR's that are created of this type.

> The client upon creating an object is giving timeouts because k8s API server takes too long to validate the resource?

To avoid this case it rejects the CRD early based on the estimation.

Reply:
Thanks for the explanation. 

And those numbers `MaxProperties=14` are magic numbers or they have a reasoning behind them? What are those number depending on?

Reply:
* The size of our Route Selector `Matches` was set to be size 8 to mirror the use in GWAPI `Matches` - https://github.com/kubernetes-sigs/gateway-api/blob/main/apis/v1/httproute_types.go#L192-L194
* The size of the `RouteSelectors` were then set as N-1 of the use in GWAPI `Rules` - https://github.com/kubernetes-sigs/gateway-api/blob/main/apis/v1/httproute_types.go#L119-L121
* The size of 14 for all of the other types that use our `RouteSelectors` is the maximum size we can set them to before we exceed the cost limit without reducing the other two limits.